tokio-util/src/io/simplex.rs

```diff
@@ -0,0 +1,233 @@
+//! Unidirectional byte-oriented channel.
+
+use bytes::Buf;
+use bytes::BytesMut;
+use futures_core::ready;
+use std::io::Error as IoError;
+use std::io::ErrorKind as IoErrorKind;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll, Waker};
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+use tokio::task::coop::poll_proceed;
+
+type IoResult<T> = Result<T, IoError>;
+
+#[derive(Debug)]
+struct Inner {
+    /// `poll_*` will return [`Poll::Pending`] if the backpressure boundary is reached
+    backpressure_boundary: usize,
+
+    /// either [`Sender`] or [`Receiver`] is closed
+    is_closed: bool,
+
+    /// Waker used to wake the [`Receiver`]
+    receiver_waker: Option<Waker>,
+
+    /// Waker used to wake the [`Sender`]
+    sender_waker: Option<Waker>,
+
+    /// Buffer used to read and write data
+    buf: BytesMut,
+}
+
+impl Inner {
+    fn with_capacity(backpressure_boundary: usize) -> Self {
+        Self {
+            backpressure_boundary,
+            is_closed: false,
+            receiver_waker: None,
+            sender_waker: None,
+            buf: BytesMut::new(),
```

Comment:
If the method is called `with_capacity`, then it shoud allocate that much memory.